google/cloud/security/common/gcp_api/_base_repository.py

```diff
@@ -0,0 +1,698 @@
+        """
+        return 'API: name=%s, versions=%s' % (self.name, self.versions)
+
+    def _init_repository(self, repository_class, version=None):
+        """Safely initialize a repository class to a property.
+
+        Args:
+          repository_class (class): The class to initialize.
+          version (str): The gcp service version for the repository.
+
+        Returns:
+          object: An instance of repository_class.
+        """
+        if not version:
+            # Use either the default version if defined or the first version
+            # returned when sorted by name.
+            version = (
+                _supported_apis.SUPPORTED_APIS.get(self.name, {})
+                .get('default_version'))
+            if not version or version not in self.gcp_services:
+                version = sorted(self.gcp_services.keys())[0]
+
+        with self._repository_lock:
+            return repository_class(gcp_service=self.gcp_services[version],
+                                    credentials=self._credentials,
+                                    rate_limiter=self._rate_limiter)
+
+
+# pylint: disable=too-many-instance-attributes, too-many-arguments
+class GCPRepository(object):
+    """Base class for GCP APIs."""
+
+    def __init__(self, gcp_service, credentials, component, entity='',
+                 num_retries=NUM_HTTP_RETRIES, key_field='project',
+                 list_key_field=None, get_key_field=None,
+                 max_results_field='maxResults', search_query_field='query',
+                 rate_limiter=None):
+        """Constructor.
+
+        Args:
+          gcp_service (object): A Resource object with methods for interacting
+              with the service.
+          credentials (object): A GoogleCredentials object
+          component (str): The subcomponent of the gcp service for this
+              repository instance. E.g. 'instances' for compute.instances().*
+              APIs
+          entity (str): The API entity returned generally by the .get() api.
+              E.g. 'instance' for compute.instances().get()
+          num_retries (int): The number of http retriable errors to retry on
+              before hard failing.
+          key_field (str): The field name representing the project to
+              query in the API.
+          list_key_field (str): Optional override of key field for calls to
+              list methods.
+          get_key_field (str): Optional override of key field for calls to
+              get methods.
+          max_results_field (str): The field name that represents the maximum
+              number of results to return in one page.
+          search_query_field (str): The field name used to filter search
+              results.
+          rate_limiter (object): A RateLimiter object to manage API quota.
+        """
+        self.gcp_service = gcp_service
+        self._credentials = credentials
+        components = component.split('.')
```

Comment:
There might be a more general way of attaching components.
Concretely, discovery.build sets__is_resource__ attribute on every nested component (ex. firewalls, networks, etc).
setattr(methodResource, '__is_resource__', True)
Therefore, we can inspect the passed service object and dynamically attach all nested repositories.

Reply:
That would be interesting to explore to further reduce the boilerplate code for the repository wrapper classes. I'm going to leave as is for now, but I agree that would be useful.